### ISSUE TITLE
Support structs in r_asm

### ISSUE THREAD

**Issue by radare** [radare/radare2#4122]
one solution would be to handle this syntax:

```
struct Foo {
int foo;
char bar;
}
```

maybe just use dword, byte, ... instead of int/char... then we should just define labels for those fields like this:

```
Foo.foo = 0
Foo.bar = 4
...
```

so in fact it will be the same as doing this:
```
Foo.foo:
.dword 0
Foo.bar:
.byte 0
```

**Comment by XVilka** [comment]
Should have support for syntax like
```
struct Foo {
...
}
```
and for 
```
struct {
...
} Foo;
```

**Comment by XVilka** [comment]
Copying text here from the GSoC Ideas page:

Unify assembly syntax among all supported architectures in radare2, implement this in an external library, and make it [fasm-compatible](http://flatassembler.net/docs.php?article=manual). This backend will perform the preprocessing assembly listings, then passing it to the frontend. It should be the one library for all available architectures.

It should support:

- Simple macro language 

```
macro <name> <args> { <body> }
```

- Preprocessor operations like “>>”, “I”, “&”, “==”, etc (should have aliases like “shr”, “and” for fasm compatibility)
- Preprocessor conditionals “if/else/switch”
- Structures with arguments, like

```
struc Unit   a,b,type
{
    .offs_l      dw a & 0xFFFF
    .cntr        db 0
    .dtype       db type
    .offs_h      dw ((b >> 16) & 0xFFFF)
    .size        = $ - .offs_l
}
```

 - Nested structures and virtual

```
pd    PStruct

; For accessing structure fields based on some register
virtual at edi
    edi.d PStruct
end virtual
mov        eax, [edi.d.Call]
```

- Support for nesting macroses and structures

**Comment by radare** [comment]
> On 14 Mar 2016, at 12:31, Anton Kochkov <notifications@github.com> wrote:
> 
> Copying text here from the GSoC Ideas page:
> 
> Unify assembly syntax among all supported architectures in radare2, implement this in an external library, and make it fasm-compatible <http://flatassembler.net/docs.php?article=manual>. This backend will perform the preprocessing assembly listings, then passing it to the frontend. It should be the one library for all available architectures.
> 
there’s no arch-specific syntax in rasm2, all the assembler directives work on all platforms. We should enhance them, by adding more assembler directives

> It should support:
> 
> Simple macro language
> macro <name> <args> { <body> }

right now the assembler is based on one liner definitions. i would proabbly move this macro task to the preprocessor, like cpp does. and in fact, spp will solve that issue (or even cpp).

i didnt implemented a preprocessor in rasm2 because rasm2 is an assembler, not a preprocessor, and the user is always free to use their favourite preprocssor. but practically nobody runs a preprocessor by hand nowadays.
> Preprocessor operations like “>>”, “I”, “&”, “==”, etc (should have aliases like “shr”, “and” for fasm compatibility)
all those binary operations for doing stuff like:

mov eax, 33*4

shuold be handled on every single assember plugin by using the r_num_math() function that will compute the value . this is really error prone, because there are many syntax from the math operations and the assembler itself that will conflict and accessing memory makes no sense to be precomputed, etc..

> Preprocessor conditionals “if/else/switch”
what about reusing spp for this? https://github.com/radare/spp or a subset of it… but i dont think inventing a new preprocessor matters.
> Structures with arguments, like
> struc Unit   a,b,type
> {
>     .offs_l      dw a & 0xFFFF
>     .cntr        db 0
>     .dtype       db type
>     .offs_h      dw ((b >> 16) & 0xFFFF)
>     .size        = $ - .offs_l
> }

isn’t this just a macro?
> Nested structures and virtual
> pd    PStruct
> 
> ; For accessing structure fields based on some register
> virtual at edi
>     edi.d PStruct
> end virtual
> mov        eax, [edi.d.Call]

im afraid this is not possible to be arch-independent, at least in the way rasm2 is done right now unless we add some kind of extra sugar to use the preprocessor. Which i would prefer because it will be more implicit and clear.
> Support for nesting macroses and structures

+1

**Comment by XVilka** [comment]
> im afraid this is not possible to be arch-independent, at least in the way rasm2 is done right now unless we add some kind of extra sugar to use the preprocessor. Which i would prefer because it will be more implicit and clear.

Yep, just smoothly integrating preprocessor is ok, the idea is just to make it unseen for the user. And I'm exactly know that 'virtual' directive is arch-independent.

**Comment by ret2libc** [comment]
This issue has been moved from radareorg/radare2 to radareorg/ideas as we are trying to clean our backlog and this issue has probably been created a long while ago. This is an effort to help contributors understand what are the actionable items they can work on, prioritize issues better and help users find active/duplicated issues more easily. If this is not an enhancement/improvement/general idea but a bug, feel free to ask for re-transfer to main repo. Thanks for your understanding and contribution with this issue.
